File: swift/Rays/Existential/Math/Quaternion.swift
import Numerics

/// A 4-component hyper-complex number used for 3D rotations and orientations.
public struct Quaternion: Hashable {

  /// The quaternion's w-component.
  public var w: Double

  /// The quaternion's x-component.
  public var x: Double

  /// The quaternion's y-component.
  public var y: Double

  /// The quaternion's z-component.
  public var z: Double

  /// Initializes a quaternion with components specified as floating-point values.
  public init(w: Double, x: Double, y: Double, z: Double) {
    self.w = w
    self.x = x
    self.y = y
    self.z = z
  }

  /// Initializes a unit quaternion from an axis-angle representation.
  ///
  /// - Parameters:
  ///   - axis: The rotation axis as a normalized 3D vector.
  ///   - angle: The magnitude of rotation.
  public init(axis: Vector3, angle: Angle) {
    let halfAngle = angle.radians * 0.5
    let s = Double.sin(halfAngle)

    self.w = Double.cos(halfAngle)
    self.x = axis.x * s
    self.y = axis.y * s
    self.z = axis.z * s
  }

  /// Initializes a unit quaternion from Euler angles.
  ///
  /// The rotations are combined following a yaw (Z), then pitch (Y), then roll (X) order.
  ///
  /// - Parameters:
  ///   - yaw: The yaw angle.
  ///   - pitch: The pitch angle.
  ///   - roll: The roll angle.
  public init(yaw: Angle, pitch: Angle, roll: Angle) {
    let cy = Double.cos(yaw.radians * 0.5)
    let sy = Double.sin(yaw.radians * 0.5)
    let cp = Double.cos(pitch.radians * 0.5)
    let sp = Double.sin(pitch.radians * 0.5)
    let cr = Double.cos(roll.radians * 0.5)
    let sr = Double.sin(roll.radians * 0.5)

    self.w = cr * cp * cy + sr * sp * sy
    self.x = sr * cp * cy - cr * sp * sy
    self.y = cr * sp * cy + sr * cp * sy
    self.z = cr * cp * sy - sr * sp * cy
  }

  /// Initializes a unit quaternion from a rotation matrix.
  public init(rotation: Matrix4) {
    let trace = rotation[0,0] + rotation[1,1] + rotation[2,2]

    if trace > 0.0 {
      var root = Double.sqrt(trace + 1.0)
      self.w = root / 2.0
      root = 0.5 / root

      self.x = (rotation[2,1] - rotation[1,2]) * root
      self.y = (rotation[0,2] - rotation[2,0]) * root
      self.z = (rotation[1,0] - rotation[0,1]) * root
    } else {
      var i = 0
      if (rotation[1,1] > rotation[0,0]) {
        i = 1
      }
      if (rotation[2,2] > rotation[i,i]) {
        i = 2
      }
      let j = (i + 1) % 3
      let k = (j + 1) % 3

      var root = Double.sqrt(rotation[i,i] - rotation[j,j] - rotation[k,k] + 1.0)

      self = .zero
      withUnsafeTemporaryAllocation(of: Double.self, capacity: 3) { (components) in
        components.initialize(repeating: 0.0)
        components[i] = 0.5 * root
        root = 0.5 / root
        components[j] = (rotation[j,i] + rotation[i,j]) * root
        components[k] = (rotation[k,i] + rotation[i,k]) * root
        
        self.w = (rotation[k,j] - rotation[j,k]) * root
        self.x = components[0]
        self.y = components[1]
        self.z = components[2]
      }
    }
  }

  /// Initializes a unit quaternion representing the rotation from one vector to another.
  ///
  /// This initializer calculates a quaternion `q` such that `q * v = u`.
  ///
  /// - Parameters:
  ///   - v: A vector.
  ///   - u: The vector obtained by applying the computed rotation.
  public init(from v: Vector3, to u: Vector3, up: Vector3 = .unitY) {
    let cos2Theta = v.dot(u)
    let vu = v.cross(u)
    let w = 1.0 + cos2Theta
    let l = Double.sqrt(w * w + vu.x * vu.x + vu.y * vu.y + vu.z * vu.z)

    self.w = w / l
    self.x = vu.x / l
    self.y = vu.y / l
    self.z = vu.z / l
  }

  /// The quaternion's magnitude (a.k.a. length or norm).
  public var magnitude: Double {
    Double.sqrt(w * w + x * x + y * y + z * z)
  }

  /// The quaternion's squared magnitude.
  public var squaredMagnitude: Double {
    w * w + x * x + y * y + z * z
  }

  /// This quaternion, normalized.
  public var normalized: Quaternion {
    let l = magnitude
    return (l != 0.0) ? (self / l) : self
  }

  /// This quaternion, inverted.
  public var inverted: Quaternion {
    let s = squaredMagnitude
    if s != 0.0 {
      let i = 1.0 / s
      return Quaternion(w: w * i, x: -x * i, y: -y * i, z: -z * i)
    } else {
      return self
    }
  }

  /// The axis-angle representation of the rotation represented by this quaternion.
  public var axisAngle: (axis: Vector3, angle: Angle) {
    get {
      let angle = Angle(radians: 2.0 * Double.acos(w))

      let d = Double.sqrt(1.0 - w * w)
      if d > 0.0 {
        return (axis: Vector3(x: x / d, y: y / d, z: z / d), angle: angle)
      } else {
        // The angle is null, so any normalized axis will do.
        return (axis: Vector3.unitX, angle: angle)
      }
    }

    set {
      self = Quaternion(axis: newValue.axis, angle: newValue.angle)
    }
  }

  /// The local yaw (i.e., z-axis rotation) element of this quaternion.
  public var yaw: Angle {
    let sycp = 2.0 * (w * z + x * y)
    let cycp = 1.0 - 2.0 * (y * y + z * z)
    return Angle(radians: Double.atan2(y: sycp, x: cycp))
  }

  /// The local pitch (i.e., y-axis rotation) element of this quaternion.
  public var pitch: Angle {
    let sp = 2.0 * (w * y + x * z)
    let rd = abs(sp) >= 1
      ? Double(signOf: sp, magnitudeOf: Double.pi / 2)
      : Double.asin(sp)
    return Angle(radians: rd)
  }

  /// The local roll (i.e., x-axis rotation) element of this quaternion.
  public var roll: Angle {
    let srcp = 2.0 * (w * x + y * z)
    let crcp = 1.0 - 2.0 * (x * x + y * y)
    return Angle(radians: Double.atan2(y: srcp, x: crcp))
  }

  /// The rotation represented by this quaternion, as a transformation matrix.
  public var transform: Matrix4 {
    get {
      let x2 = x + x
      let y2 = y + y
      let z2 = z + z

      let lengthSquared = w * w + x * x + y * y + z * z
      let wx = w * x2 / lengthSquared
      let wy = w * y2 / lengthSquared
      let wz = w * z2 / lengthSquared
      let xx = x * x2 / lengthSquared
      let xy = x * y2 / lengthSquared
      let xz = x * z2 / lengthSquared
      let yy = y * y2 / lengthSquared
      let yz = y * z2 / lengthSquared
      let zz = z * z2 / lengthSquared

      var result = Matrix4.zero

      result[0,0] = 1.0 - (yy + zz)
      result[1,0] = xy + wz
      result[2,0] = xz - wy
      result[0,1] = xy - wz
      result[1,1] = 1.0 - (xx + zz)
      result[2,1] = yz + wx
      result[0,2] = xz + wy
      result[1,2] = yz - wx
      result[2,2] = 1.0 - (xx + yy)
      result[3,3] = 1.0

      return result
    }

    set {
      self = Quaternion(rotation: newValue)
    }
  }

  /// Returns the dot (a.k.a. scalar) product of `self` with `other`.
  public func dot(_ other: Quaternion) -> Double {
    w * other.w + x * other.x + y * other.y + z * other.z
  }

  /// Computes the normalized linear interpolation (nlerp) between this quaternion and another.
  ///
  /// - Parameters:
  ///   - other: The other quaternion between which compute the interpolation.
  ///   - distance: The distance from this quaternion to the other, in the range `[0, 1]`.
  public func nlrep(with other: Quaternion, distance: Double) -> Quaternion {
    (self + (other - self) * distance).normalized
  }

  /// Computes the spherical linear interpolation (slerp) between this quaternion and another.
  ///
  /// - Parameters:
  ///   - other: The other quaternion between which compute the interpolation.
  ///   - distance: The distance from this quaternion to the other, in the range `[0, 1]`.
  public func slerp(with other: Quaternion, distance: Double) -> Quaternion {
    let cosHalfTheta = dot(other)
    if cosHalfTheta < 0.95 {
      // Standard case (slerp)
      let sinHalfTheta = Double.sqrt(1.0 - cosHalfTheta * cosHalfTheta)
      let angle = Double.atan2(y: sinHalfTheta, x: cosHalfTheta)
      let coef0 = Double.sin((1.0 - distance) * angle) / sinHalfTheta
      let coef2 = Double.sin(distance * angle) / sinHalfTheta
      return self * coef0 + other * coef2
    } else {
      // There are two situations here:
      // - `self` and `other` are very close, so a linear interpolation is safe.
      // - `self` and `other` are almost inverse of each other, thus there is an infinite number of
      //   possible interpolations. Since we don't have a strategy to solve this situation, we may
      //   as well get away with a linear interpolation.
      return nlrep(with: other, distance: distance)
    }
  }

  /// Returns the component-wise addition of two quaternions.
  public static func + (lhs: Quaternion, rhs: Quaternion) -> Quaternion {
    .init(w: lhs.w + rhs.w, x: lhs.x + rhs.x, y: lhs.y + rhs.y, z: lhs.z + rhs.z)
  }

  /// Computes the component-wise addition of two quaternions and stores the result in `lhs`.
  public static func += (lhs: inout Quaternion, rhs: Quaternion) {
    lhs.w += rhs.w
    lhs.x += rhs.x
    lhs.y += rhs.y
    lhs.z += rhs.z
  }

  /// Returns the component-wise subtraction of two quaternions.
  public static func - (lhs: Quaternion, rhs: Quaternion) -> Quaternion {
    .init(w: lhs.w - rhs.w, x: lhs.x - rhs.x, y: lhs.y - rhs.y, z: lhs.z - rhs.z)
  }

  /// Computes the component-wise subtraction of two quaternions and stores the result in `lhs`.
  public static func -= (lhs: inout Quaternion, rhs: Quaternion) {
    lhs.w -= rhs.w
    lhs.x -= rhs.x
    lhs.y -= rhs.y
    lhs.z -= rhs.z
  }

  /// Returns the multiplication of two quaternions.
  ///
  /// If `q` and `r` are unit quaternions, then their multiplication is the composition (a.k.a.
  /// concatenation) of the rotation they represent.
  ///
  /// This operation is not commutative. There are pairs of quaternions for which `q * r != r * q`.
  public static func * (q: Quaternion, r: Quaternion) -> Quaternion {
    .init(
      w: q.w * r.w - q.x * r.x - q.y * r.y - q.z * r.z,
      x: q.w * r.x + q.x * r.w + q.y * r.z - q.z * r.y,
      y: q.w * r.y + q.y * r.w + q.x * r.z - q.z * r.x,
      z: q.w * r.z + q.z * r.w + q.x * r.y - q.y * r.x)
  }

  /// Returns the multiplication of two quaternions and stores the result in `lhs`.
  public static func *= (lhs: inout Quaternion, rhs: Quaternion) {
    lhs.w *= rhs.w
    lhs.x *= rhs.x
    lhs.y *= rhs.y
    lhs.z *= rhs.z
  }

  /// Returns the rotation of a vector by a quaternion.
  public static func * (v: Vector3, q: Quaternion) -> Vector3 {
    // nVidia SDK implementation
    let qvec = Vector3(x: q.x, y: q.y, z: q.z)
    var uv = qvec.cross(v)
    var uuv = qvec.cross(uv)
    uv = uv * (2.0 * q.w)
    uuv = uuv * 2.0
    return v + uv + uuv
  }

  /// Returns the multiplication of a quaternion by a scalar.
  public static func * (lhs: Quaternion, rhs: Double) -> Quaternion {
    .init(w: lhs.w * rhs, x: lhs.x * rhs, y: lhs.y * rhs, z: lhs.z * rhs)
  }

  /// Computes the multiplication of a quaternion by a scalar and stores the result in `lhs`.
  public static func *= (lhs: inout Quaternion, rhs: Double) {
    lhs.w *= rhs
    lhs.x *= rhs
    lhs.y *= rhs
    lhs.z *= rhs
  }

  /// Returns the division of a quaternion by a scalar.
  public static func / (lhs: Quaternion, rhs: Double) -> Quaternion {
    .init(w: lhs.w / rhs, x: lhs.x / rhs, y: lhs.y / rhs, z: lhs.z / rhs)
  }

  /// Computes the division of a quaternion by a scalar and stores the result in `lhs`.
  ///
  /// - Parameters:
  ///   - lhs: The quaternion to divide.
  ///   - rhs: A scalar value.
  public static func /= (lhs: inout Quaternion, rhs: Double) {
    lhs.w /= rhs
    lhs.x /= rhs
    lhs.y /= rhs
    lhs.z /= rhs
  }

  /// A quaternion whose 4 components are zero.
  public static let zero = Quaternion(w: 0.0, x: 0.0, y: 0.0, z: 0.0)

  /// A unit quaternion representing the identity rotation.
  public static let identity = Quaternion(w: 1.0, x: 0.0, y: 0.0, z: 0.0)

}

extension Quaternion: CustomStringConvertible {

  public var description: String {
    "(roll: \(roll.degrees), pitch: \(pitch.degrees), yaw: \(yaw.degrees))"
  }

}

File: scala/src/main/scala/rays/inheritance/math/Quaternion.scala
package rays.inheritance.math

import scala.math.{abs, acos, asin, atan2, sqrt, cos, sin, Pi}

/** A 4-component hyper-complex number used for 3D rotations and orientations.
  *
  * @param w The quaternion's w-component.
  * @param x The quaternion's x-component.
  * @param y The quaternion's y-component.
  * @param z The quaternion's z-component.
  */
final case class Quaternion(w: Double, x: Double, y: Double, z: Double):

  /** The quaternion's magnitude (a.k.a. length or norm). */
  def magnitude: Double =
    sqrt(w * w + x * x + y * y + z * z)

  /** The quaternion's squared magnitude.
    *
    * Use this property rather than `magnitude` if you do not need the exact magnitude of the
    * quaternion, but just want know if it is `0` or if it is longer than another quaternion's.
    */
  def squaredMagnitude: Double =
    w * w + x * x + y * y + z * z

  /** The quaternion, normalized. */
  def normalized: Quaternion =
    val l = magnitude
    if l != 0.0 then this / l else this

  /** The quaternion, inverted. */
  def inverted: Quaternion =
    val s = squaredMagnitude
    if s != 0.0 then
      val i = 1.0 / s
      Quaternion(w * i, -x * i, -y * i, -z * i)
    else
      this

  /** The axis-angle representation of the rotation represented by this quaternion. */
  def axisAngle: (Vector3, Angle) =
    val a = Angle.radians(2.0 * acos(w))
    val d = sqrt(1.0 - w * w)
    if d > 0.0 then
      (Vector3(x / d, y / d, z / d), a)
    else
      // The angle is null, so any normalized axis will do.
      return (Vector3.unitX, a)

  /** The local yaw (i.e., z-axis rotation) element of this quaternion. */
  def yaw: Angle =
    val sycp = 2.0 * (w * z + x * y)
    val cycp = 1.0 - 2.0 * (y * y + z * z)
    Angle.radians(atan2(sycp, cycp))

  /** The local pitch (i.e., y-axis rotation) element of this quaternion. */
  def pitch: Angle =
    val sp = 2.0 * (w * y + x * z)
    if abs(sp) >= 1 then
      Angle.radians(if sp > 0 then (Pi / 2) else -(Pi / 2))
    else
      Angle.radians(asin(sp))

  /** The local roll (i.e., x-axis rotation) element of this quaternion. */
  def roll: Angle =
    val srcp = 2.0 * (w * x + y * z)
    val crcp = 1.0 - 2.0 * (x * x + y * y)
    Angle.radians(atan2(srcp, crcp))

  /** Returns the dot (a.k.a. scalar) product of `this` with `that`. */
  def dot(that: Quaternion): Double =
    w * that.w + x * that.x + y * that.y + z * that.z

  /** Computes the normalized linear interpolation (nlerp) between this quaternion and another.
    *
    * @param that The other quaternion between which compute the interpolation.
    * @param distance The distance from this quaternion to the other, in the range `[0, 1]`.
    */
  def nlrep(that: Quaternion, distance: Double): Quaternion =
    (this + (that - this) * distance).normalized

  /** Computes the spherical linear interpolation (slerp) between this quaternion and another.
    *
    * @param other: The other quaternion between which compute the interpolation.
    * @param distance: The distance from this quaternion to the other, in the range `[0, 1]`.
    */
  def slerp(that: Quaternion, distance: Double): Quaternion =
    val cosHalfTheta = dot(that)
    if cosHalfTheta < 0.95 then
      // Standard case (slerp)
      val sinHalfTheta = sqrt(1.0 - cosHalfTheta * cosHalfTheta)
      val angle = atan2(sinHalfTheta, cosHalfTheta)
      val coef0 = sin((1.0 - distance) * angle) / sinHalfTheta
      val coef2 = sin(distance * angle) / sinHalfTheta
      this * coef0 + that * coef2
    else
      // There are two situations here:
      // - `self` and `other` are very close, so a linear interpolation is safe.
      // - `self` and `other` are almost inverse of each other, thus there is an infinite number of
      //   possible interpolations. Since we don't have a strategy to solve this situation, we may
      //   as well get away with a linear interpolation.
      nlrep(that, distance)

  /** Returns the component-wise addition of two quaternions. */
  def + (that: Quaternion): Quaternion =
    Quaternion(w + that.w, x + that.x, y + that.y, z + that.z)

  /** Returns the component-wise subtraction of two quaternions. */
  def - (that: Quaternion): Quaternion =
    Quaternion(w - that.w, x - that.x, y - that.y, z - that.z)

  /** Returns the multiplication of two quaternions.
    *
    * If `this` and `that` are unit quaternions, then their multiplication is the composition (a.k.a.
    * concatenation) of the rotation they represent.
    *
    * This operation is not commutative. There exist pairs of quaternions `q`, `r` such that
    * `q * r != r * q`.
    */
  def * (that: Quaternion): Quaternion =
    Quaternion(
      w * that.w - x * that.x - y * that.y - z * that.z,
      w * that.x + x * that.w + y * that.z - z * that.y,
      w * that.y + y * that.w + x * that.z - z * that.x,
      w * that.z + z * that.w + x * that.y - y * that.x)

  /** Returns the multiplication of a quaternion by a scalar. */
  def * (that: Double): Quaternion =
    Quaternion(w * that, x * that, y * that, z * that)

  /** Returns the division of a quaternion by a scalar. */
  def / (that: Double): Quaternion =
    Quaternion(w / that, x / that, y / that, z / that)

  /** A string representation of `this`. */
  override def toString: String =
    s"Rotation(yaw: ${yaw}, pitch: ${pitch}, roll: ${roll})"

object Quaternion:

  /** A quaternion whose 4 components are zero. */
  final val zero = Quaternion(0.0, 0.0, 0.0, 0.0)

  /** A unit quaternion representing the identity rotation. */
  final val identity = Quaternion(1.0, 0.0, 0.0, 0.0)

  /** Returns a unit quaternion from an axis-angle representation.
   *
   * @param axis The rotation axis as a normalized 3D vector.
   * @param angle: The magnitude of rotation.
   */
  def fromAxisAngle(axis: Vector3, angle: Angle): Quaternion =
    val h = angle.radians * 0.5
    val s = sin(h)
    Quaternion(cos(h), axis.x * s, axis.y * s, axis.z * s)

  /** Returns a unit quaternion from Euler angles.
    *
    * @param yaw The yaw angle.
    * @param pitch The pitch angle.
    * @param roll The roll angle.
    */
  def fromEuler(yaw: Angle, pitch: Angle, roll: Angle): Quaternion =
    val cy = cos(yaw.radians * 0.5)
    val sy = sin(yaw.radians * 0.5)
    val cp = cos(pitch.radians * 0.5)
    val sp = sin(pitch.radians * 0.5)
    val cr = cos(roll.radians * 0.5)
    val sr = sin(roll.radians * 0.5)
    Quaternion(
      cr * cp * cy + sr * sp * sy,
      sr * cp * cy - cr * sp * sy,
      cr * sp * cy + sr * cp * sy,
      cr * cp * sy - sr * sp * cy)

  /** Initializes a unit quaternion representing the rotation from one vector to another.
    *
    * @param source A vector.
    * @param target The vector obtained by applying the computed rotation.
    */
  def fromRotation(source: Vector3, target: Vector3, up: Vector3 = Vector3.unitY): Quaternion =
    val cos2Theta = source.dot(target)
    val vu = source.cross(target)
    val w = 1.0 + cos2Theta
    val l = sqrt(w * w + vu.x * vu.x + vu.y * vu.y + vu.z * vu.z)
    Quaternion(w / l, vu.x / l, vu.y / l, vu.z / l)

end Quaternion
